tests/test_base_task.py

```diff
@@ -0,0 +1,103 @@
+import pandas as pd
+import pytest
+
+from hrqb.base import PandasPickleTarget, PandasPickleTask, QuickbaseTableTarget
+
+
+def test_pandas_pickle_task_gives_pandas_pickle_target(pandas_pickle_task):
+    target = pandas_pickle_task.target()
+    assert isinstance(target, PandasPickleTarget)
+    assert target.path == pandas_pickle_task.path
+    assert target.table_name == pandas_pickle_task.table_name
+
+
+def test_pandas_pickle_task_output_path_is_target_path(pandas_pickle_task):
+    assert pandas_pickle_task.output().path == pandas_pickle_task.target().path
+
+
+def test_quickbase_upsert_task_gives_quickbase_table_target(quickbase_upsert_task):
+    target = quickbase_upsert_task.target()
+    assert isinstance(target, QuickbaseTableTarget)
+    assert target.path == quickbase_upsert_task.path
+    assert target.table_name == quickbase_upsert_task.table_name
+
+
+def test_quickbase_upsert_task_output_path_is_target_path(quickbase_upsert_task):
+    assert quickbase_upsert_task.output().path == quickbase_upsert_task.target().path
+
+
+def test_luigi_task_incomplete_when_target_not_exists(tmpdir, pandas_pickle_task):
+    assert not pandas_pickle_task.complete()
+
+
+def test_luigi_task_complete_when_target_exists(tmpdir, pandas_pickle_task):
+    with open(pandas_pickle_task.path, "a"):
+        assert pandas_pickle_task.complete()
```

Comment:
What makes a target "exist"? 🤔 How do these tests operate / differ?

Reply:
It's a sublte difference between a Target and a Task.

A Target can "exist" based on its `.exists()` method.  Usually, this is looking for a file on disk, but in the case of a database table, it might be that DB table exists.

A Task is "complete" when all of its own Targets "exist".  Here is the [source code for `Task.complete`](https://github.com/spotify/luigi/blob/master/luigi/task.py#L588-L603).  

I'm a little unsure at the moment how or when a Task could have multiple Targets, but I do think it's conceptually possible.

Scratch that, [`HRQBPipelineTask` is an in this project](https://github.com/MITLibraries/hrqb-client/pull/7/files#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R69-R70).

It doesn't do much yet, but a more realistic example of a pipeline Task will look like:
```python
class LookupTablesPipeline(HRQBPipelineTask):
    def requires(self):
        yield LoadJobTitles()
        yield LoadPositionTitles()
```

It's not very obvious, I'll give you that, but I think this Task is "complete" when all the tasks it requires and has yielded the result of, are themselves complete.  They are like the output of this special `luigi.WrapperTask`.  Probably a bad example, but suffice to say that a Task can have multiple outputs, and it's "complete" when they all "exist".

Good questions though!  As @ehanson8 rightly pointed out: it's a bit opaque at the moment, but some PRs in the works will click these things into place and I think their linkages will make more sense.

Reply:
But I'm rereading the line highlighted, and your question...

This test is confirming that when the target of a Task exists, the Task itself is complete. It simulates this behavior by manually creating the file via the `with open(...)` and therefore suddenly the Task is complete.

Reply:
Thank you for this additional information!